### PR TITLE
Develop → Main

### PR DESCRIPTION
**Pull request by ksookdhew**
## Description 
Merging Dev into Main
 
## Screenshots/GIFs 
Constructor
Constructor Standing
Driver
Driver Standing
Results
Race Info
Racing
Race result

## Checklist 
Please review and check the following before submitting this pull request: 
- [x] The code follows the project's coding standards. 
- [x] 100% Unit test coverage for View Models 
- [x] All warnings Resolved.

### CODE REVIEW

File: F1/F1/Constructor/ConstructorViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+//  Created by Kaitlyn Sookdhew on 2024/04/02.
+//
+
+import Foundation
+
+class ConstructorViewModel {
+
+    // MARK: Variables
+    private var repository: ConstructorRepositoryType?
+    private weak var delegate: ViewModelDelegate?
```
Comment (simekanidvt) [inline F1/F1/Constructor/ConstructorViewModel.swift:14]:
```suggestion
     private weak var delegate: ViewModelDelegate?
    private var repository: ConstructorRepositoryType?
    private var constructorResults: [Race]?
```
Nitpick: Try keep variables of same type grouped together

File: F1/F1/Constructor/ConstructorViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private var constructorResults: [Race]?
+
+    init(repository: ConstructorRepositoryType,
+         delegate: ViewModelDelegate) {
```
Comment (simekanidvt) [inline F1/F1/Constructor/ConstructorViewModel.swift:18]: This can fit on one line

File: F1/F1/Driver/DriverViewController.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/02.
+//
+
+import UIKit
+
+class DriverViewController: UIViewController {
+
+    @IBOutlet weak private var tableView: UITableView!
```
Comment (simekanidvt) [inline F1/F1/Driver/DriverViewController.swift:12]: Group variables together. will just look neater

File: F1/F1/Driver/DriverViewController.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+extension  DriverViewController: ViewModelDelegate {
+
+    func reloadView() {
+        tableView.reloadData()
+    }
+
+    func show(error: String) {
+    }
```
Comment (simekanidvt) [inline F1/F1/Driver/DriverViewController.swift:90]: Are you leaving this empty?

File: F1/F1/Driver/DriverViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+class DriverViewModel {
+
+    // MARK: Variables
+    private var repository: DriverRepositoryType?
+    private weak var delegate: ViewModelDelegate?
+    private var driverResults: [Race]?
+
+    init(repository: DriverRepositoryType,
+         delegate: ViewModelDelegate) {
```
Comment (simekanidvt) [inline F1/F1/Driver/DriverViewModel.swift:18]: single line

File: F1/F1/Driver/DriverViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        driverResults?.count ?? 0
+    }
+
+    // MARK: Functions
+    func result(atIndex: Int) -> Race? {
+        driverResults?[atIndex] ?? nil
+    }
+
+    func laptime(index: Int) -> String {
+        if let resultTime = driverResults?[index].results.first?.time {
```
Comment (simekanidvt) [inline F1/F1/Driver/DriverViewModel.swift:34]: Try clean this up with guard let. The nested if statements are a bit messy

File: F1/F1/Login/LoginViewController.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    @IBOutlet weak var errorMsg: UILabel!
+    private lazy var viewModel = LoginViewModel()
+
+    override func viewDidLoad() {
+        errorMsg.isHidden = true
+        super.viewDidLoad()
+    }
+
+    @IBAction func loginAction(_ sender: Any) {
+        if viewModel.validDetails(givenUsername: username.text ?? "", givenPassword: password.text ?? "") {
```
Comment (simekanidvt) [inline F1/F1/Login/LoginViewController.swift:23]: please move this logic to view model and use delegate methods

File: F1/F1/Race/RaceViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+//
+
+import Foundation
+
+class RaceViewModel {
+
+    // MARK: Variables
+    private var repository: RaceRepositoryType?
+    private weak var delegate: ViewModelDelegate?
+    private(set) var allRaces: [RaceInfo] = []
```
Comment (simekanidvt) [inline F1/F1/Race/RaceViewModel.swift:15]: Please group variables

File: F1/F1/Race/RaceViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                sortedRaceSession.append(RaceSessionDetail(date: secondPractice.date, time: secondPractice.time, type: .sprintQualifying))
+            }
+        } else if let secondPractice = race?.secondPractice {
+            sortedRaceSession.append(RaceSessionDetail(date: secondPractice.date, time: secondPractice.time, type: .practice2))
+        }
+
+        if let firstPractice = race?.firstPractice {
+            sortedRaceSession.append(RaceSessionDetail(date: firstPractice.date, time: firstPractice.time, type: .practice1))
+        }
+
```
Comment (simekanidvt) [inline F1/F1/Race/RaceViewModel.swift:71]: Unnecessary line break

File: F1/F1/Race/Repository/RaceRepository.swift [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/03.
+//
+
+import Foundation
+typealias RaceResults = (Result< Racing, APIError>) -> Void
+
+protocol RaceRepositoryType: AnyObject {
+    func fetchRaceResults(completion: @escaping(RaceResults))
```
Comment (simekanidvt) [inline F1/F1/Race/Repository/RaceRepository.swift:12]:
```suggestion
    func fetchRaceResults(completion: @escaping RaceResults)
```

File: F1/F1/Config/AppDelegate.swift [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/02.
+//
+
+import UIKit
+
+@main
+class AppDelegate: UIResponder, UIApplicationDelegate {
+
+
```
Comment (kieran-woodrow) [inline F1/F1/Config/AppDelegate.swift:13]: unnecessary new lines

File: F1/F1/Config/ViewController.swift [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+import UIKit
+
+class ViewController: UIViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view.
+    }
+
```
Comment (kieran-woodrow) [inline F1/F1/Config/ViewController.swift:16]: unnecessary new lines

Comment (kieran-woodrow) [reply F1/F1/Config/ViewController.swift:16]: also is this file being used? It doesn't look like it? So you can remove it please

File: F1/F1/Constructor/Constructor.swift [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    enum CodingKeys: String, CodingKey {
+        case constructor = "MRData"
+    }
+}
+
+// MARK: - ConstructorResponse
+struct ConstructorResponse: Codable {
+    let series: String
+    let url: String
+    let limit, offset, total: String
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/Constructor.swift:23]: any reason some of these strings are on their own line but then the rest of the other strings are on one line? Coose a standard and stick to it
 
 let series: String
    let url: String
    let limit, offset, total: String

Comment (kieran-woodrow) [reply F1/F1/Constructor/Constructor.swift:23]: do the same everywhere please

File: F1/F1/Constructor/Constructor.swift [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        case constructorID = "constructorId"
+        case constructors = "Constructors"
+    }
+}
+
+// MARK: - Constructor
+struct Constructor: Codable {
+    let constructorID: String
+    let url: String
+    let name, nationality: String
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/Constructor.swift:47]: same here

File: F1/F1/Constructor/ConstructorRepository.swift [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+typealias ConstructorResults = (Result<RacingResults, APIError>) -> Void
+
+protocol ConstructorRepositoryType: AnyObject {
+    func fetchConstructorResults(constructor: String, completion: @escaping ConstructorResults)
+}
+
+class ConstructorRepository: ConstructorRepositoryType {
+
+    func fetchConstructorResults(constructor: String, completion: @escaping ConstructorResults) {
+        let url = "https://ergast.com/api/f1/current/constructors/\(constructor)/results.JSON"
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/ConstructorRepository.swift:19]: Please add this to the constants file

File: F1/F1/Constructor/ConstructorViewController.swift [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+                                                      delegate: self)
+    @IBOutlet weak private var tableView: UITableView!
+    @IBOutlet weak private var constructorImage: UIImageView!
+    @IBOutlet weak private var constructorName: UILabel!
+    @IBOutlet weak private var constructorNationality: UILabel!
+    @IBOutlet weak private var constructorDriver: UILabel!
+    @IBOutlet weak private var currentWins: UILabel!
+    @IBOutlet weak private var currentPoints: UILabel!
+    @IBOutlet weak private var currentPosition: UILabel!
+    var constructor: ConstructorStanding?
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/ConstructorViewController.swift:21]: Put a new line after outlets and a mark for variables

File: F1/F1/Constructor/ConstructorViewController.swift [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+        tableView.dataSource = self
+        tableView.register(ConstructorTableViewCell.nib(),
+                           forCellReuseIdentifier: ConstructorTableViewCell.identifier)
+        tableView.register(UINib(nibName: ConstructorResultHeader.identifier, bundle: nil),
+                           forHeaderFooterViewReuseIdentifier: ConstructorResultHeader.identifier)
+    }
+
+    private func setupView() {
+        constructorImage.image = UIImage(named: viewModel.imageName(constructorId:constructor?.constructor.constructorID))
+        constructorName.text = constructor?.constructor.name
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/ConstructorViewController.swift:42]: A lot of this can be in the viewModel as computed properties

File: F1/F1/Constructor/ConstructorViewController.swift [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    func show(error: String) {
+
+    }
+}
+
+// MARK: - UITableViewDelegate and UITableViewDataSource
+extension ConstructorViewController: UITableViewDelegate, UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/ConstructorViewController.swift:65]: return statements

File: F1/F1/Constructor/ConstructorViewController.swift [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        return viewModel.resultsCount
+    }
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/ConstructorViewController.swift:67]: new lines after these functions end and start

File: F1/F1/Constructor/constructorTableViewCell.swift [hunk 20]
```diff
@@ -0,0 +1,51 @@
+//
+//  ConstructorTableViewCell.swift
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/08.
+//
+
+import UIKit
+
+class ConstructorTableViewCell: UITableViewCell {
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/constructorTableViewCell.swift:10]: missing marks

File: F1/F1/Constructor/constructorTableViewCell.swift [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+    @IBOutlet weak private var round: UILabel!
+    @IBOutlet weak private var race: UILabel!
+    @IBOutlet weak private var firstDriver: UILabel!
+    @IBOutlet weak private var firstPoints: UILabel!
+    @IBOutlet weak private var firstPosition: UILabel!
+    @IBOutlet weak private var secondDriver: UILabel!
+    @IBOutlet weak private var secondPoints: UILabel!
+    @IBOutlet weak private var secondPosition: UILabel!
+    
+    static let identifier = Identifiers.constructorTableViewCell
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/constructorTableViewCell.swift:21]: can add this to constants and remove static let identifier = Identifiers.constructorTableViewCell

File: F1/F1/Constructor/constructorTableViewCell.swift [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        firstDriver.layer.cornerRadius = 8
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/constructorTableViewCell.swift:24]: can try do this in storyboards

File: F1/F1/Constructor/constructorTableViewCell.swift [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    func populateWith(result: Race) {
+        round.text = result.round
+        race.text = result.circuit.location.locality
+        firstDriver.text = result.results.first?.driver.code
+        firstPoints.text = result.results.first?.points
+        firstPosition.text = result.results.first?.position
+
+        if result.results.count > 1 {
```
Comment (kieran-woodrow) [inline F1/F1/Constructor/constructorTableViewCell.swift:37]: can rather test for if if !result.results.count.isEmpty(). Avoid used count > 1 if you can to make your code look a lot more "professional"

File: F1/F1/Driver/DriverResultHeader.swift [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+//  DriverResultHeader.swift
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/09.
+//
+
+import UIKit
+
+class DriverResultHeader: UITableViewHeaderFooterView {
+    static let identifier = Identifiers.driverResultsIdentifier
```
Comment (kieran-woodrow) [inline F1/F1/Driver/DriverResultHeader.swift:11]: please add to constants file

File: F1/F1/Driver/DriverViewController.swift [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+        tableView.delegate = self
+        tableView.dataSource = self
+        tableView.register(DriverTableViewCell.nib(),
+                           forCellReuseIdentifier: DriverTableViewCell.identifier)
+        tableView.register(DriverResultHeader.nib(),
+                           forHeaderFooterViewReuseIdentifier: DriverResultHeader.identifier)
+    }
+
+    private func setupView() {
+        driverImg.image = UIImage(named: viewModel.imageName(driverCode: driver?.driver.code))
```
Comment (kieran-woodrow) [inline F1/F1/Driver/DriverViewController.swift:42]: can add a lot of these to viewModel and call these as computed properties

File: F1/F1/Driver/DriverViewController.swift [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+extension DriverViewController: UITableViewDelegate, UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return viewModel.resultsCount
+    }
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 42.0
+    }
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: DriverTableViewCell.identifier)
+                as? DriverTableViewCell
```
[the remark this answers is not in the document]
Comment (kieran-woodrow) [reply F1/F1/Driver/DriverViewController.swift:63]: and return statements

File: F1/F1/Driver/Repository/Driver.swift [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+    enum CodingKeys: String, CodingKey {
+        case driver = "MRData"
+    }
+}
+
+// MARK: - DriverResponse
+struct DriverResponse: Codable {
+    let series: String
+    let url: String
+    let limit, offset, total: String
```
Comment (kieran-woodrow) [inline F1/F1/Driver/Repository/Driver.swift:23]: same as previous comment

File: F1/F1/Driver/Repository/DriverRepository.swift [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+
+protocol DriverRepositoryType: AnyObject {
+
+    func fetchDriverResults(driver: String, completion: @escaping DriverResults)
+}
+
+class DriverRepository: DriverRepositoryType {
+
+    func fetchDriverResults(driver: String, completion: @escaping DriverResults) {
+        let url = "https://ergast.com/api/f1/current/drivers/\(driver)/results.JSON"
```
Comment (kieran-woodrow) [inline F1/F1/Driver/Repository/DriverRepository.swift:20]: can be added to constants

File: F1/F1/Driver/driverTableViewCell.swift [hunk 29]
```diff
@@ -0,0 +1,36 @@
+//
+//  DriverTableViewCell.swift
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/08.
+//
+
+import UIKit
+
+class DriverTableViewCell: UITableViewCell {
```
Comment (kieran-woodrow) [inline F1/F1/Driver/driverTableViewCell.swift:10]: marks

File: F1/F1/Driver/driverTableViewCell.swift [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+
+    @IBOutlet weak private var round: UILabel!
+    @IBOutlet weak private var race: UILabel!
+    @IBOutlet weak private var time: UILabel!
+    @IBOutlet weak private var points: UILabel!
+    @IBOutlet weak private var position: UILabel!
+
+    static let identifier = Identifiers.driverTableViewCell
```
Comment (kieran-woodrow) [inline F1/F1/Driver/driverTableViewCell.swift:18]: constants

File: F1/F1/Driver/driverTableViewCell.swift [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        time.layer.cornerRadius = 8
```
Comment (kieran-woodrow) [inline F1/F1/Driver/driverTableViewCell.swift:21]: storyboard

File: F1/F1/Driver/driverTableViewCell.swift [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+        time.layer.masksToBounds = true
+    }
+
+    func populateWith(result: Race, lapTime: String) {
+        round.text = result.round
+        race.text = result.circuit.location.locality
+        time.text = lapTime
+        points.text = result.results[0].points
```
Comment (kieran-woodrow) [inline F1/F1/Driver/driverTableViewCell.swift:29]: .first instead of [0]

File: F1/F1/Login/LoginRepository.swift [hunk 33]
```diff
@@ -0,0 +1,24 @@
+//
+//  LoginRepository.swift
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/10.
+//
+
+import Foundation
```
Comment (kieran-woodrow) [inline F1/F1/Login/LoginRepository.swift:8]: marks, new lines

File: F1/F1/Login/LoginViewController.swift (hunk 7 above)
Comment (kieran-woodrow) [reply F1/F1/Login/LoginViewController.swift:23]: marks

File: F1/F1/Login/LoginViewController.swift [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+
+class LoginViewController: UIViewController {
+
+    @IBOutlet weak var username: UITextField!
+    @IBOutlet weak var password: UITextField!
+    @IBOutlet weak var errorMsg: UILabel!
+    private lazy var viewModel = LoginViewModel()
+
+    override func viewDidLoad() {
+        errorMsg.isHidden = true
```
Comment (kieran-woodrow) [inline F1/F1/Login/LoginViewController.swift:18]: abbreviation

File: F1/F1/Login/LoginViewController.swift [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+        super.viewDidLoad()
+    }
+
+    @IBAction func loginAction(_ sender: Any) {
+        if viewModel.validDetails(givenUsername: username.text ?? "", givenPassword: password.text ?? "") {
+            errorMsg.isHidden = true
+            performSegue(withIdentifier: "showTabSegue", sender: self)
```
Comment (kieran-woodrow) [inline F1/F1/Login/LoginViewController.swift:25]: show segue to constants

File: F1/F1/Race/RaceScheduleTableViewCell.swift [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+    // MARK: IBOutlet
+    @IBOutlet weak private var raceDay: UILabel!
+    @IBOutlet weak private var raceMonth: UILabel!
+    @IBOutlet weak private var session: UILabel!
+    @IBOutlet weak private var time: UILabel!
+
+    // MARK: Functions
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        contentView.layer.cornerRadius = 16
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceScheduleTableViewCell.swift:21]: storyboard

File: F1/F1/Race/RaceViewController.swift [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+    @IBOutlet weak private var circuit: UILabel!
+    @IBOutlet weak private var raceLocation: UILabel!
+    @IBOutlet weak private var tableView: UITableView!
+
+    // MARK: Variables
+    var raceInfo: RaceInfo?
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        viewModel.race = raceInfo
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceViewController.swift:24]: can this not be done in the viewModel?

File: F1/F1/Race/RaceViewController.swift [hunk 38]
```diff
@@ -0,0 +1,87 @@
...
+        setupView()
+        setupTableView()
+    }
+
+    private func setupTableView() {
+        tableView.delegate = self
+        tableView.dataSource = self
+        tableView.register(RaceScheduleTableViewCell.nib(), forCellReuseIdentifier: Identifiers.raceScheduleIndentifier)
+        tableView.sectionHeaderTopPadding = 0.0
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceViewController.swift:33]: this seems irrelevant to have 0.0 padding?

File: F1/F1/Race/RaceViewController.swift [hunk 39]
```diff
@@ -0,0 +1,87 @@
...
+
+// MARK: - UITableViewDelegate and UITableViewDataSource
+extension RaceViewController: UITableViewDelegate, UITableViewDataSource {
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        viewModel.scheduleCount
+    }
+
+   func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+       return 1
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceViewController.swift:62]: return statements

File: F1/F1/Race/RaceViewController.swift [hunk 40]
```diff
@@ -0,0 +1,87 @@
...
+   }
+   func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+       return 10
+   }
+
+   func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+       return 69.0
+    }
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceViewController.swift:70]: new line

File: F1/F1/Race/RaceViewModel.swift [hunk 41]
```diff
@@ -0,0 +1,110 @@
...
+    func imageName(circuitCode: String?) -> String {
+        "\(circuitCode ?? "").png"
+    }
+
+    func raceSession(atIndex: Int) -> RaceSessionDetail {
+        sortedRaceSession[atIndex]
+    }
+
+    func sessionDate(date: String) -> DateComponents {
+        let dateFormatter = DateFormatter()
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceViewModel.swift:87]: consider hving a dateFormatter class for reusibility

File: F1/F1/Race/Repository/Racing.swift [hunk 42]
```diff
@@ -0,0 +1,89 @@
...
+    enum CodingKeys: String, CodingKey {
+        case race = "MRData"
+    }
+}
+
+// MARK: - RaceDescriptor
+struct RaceDescriptor: Codable {
+    let series: String
+    let url: String
+    let limit, offset, total: String
```
Comment (kieran-woodrow) [inline F1/F1/Race/Repository/Racing.swift:23]: same as before

File: F1/F1/Repositories/F1Repository.swift [hunk 43]
```diff
@@ -0,0 +1,11 @@
+//
+//  F1Repository.swift
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/02.
```
Comment (kieran-woodrow) [inline F1/F1/Repositories/F1Repository.swift:5]: empty file?

File: F1/F1/Results/AllResultsTableViewCell.swift [hunk 44]
```diff
@@ -0,0 +1,56 @@
...
+        setUpCell()
+    }
+
+    func setUpCell() {
+        contentView.layer.cornerRadius = 10
+        contentView.layer.masksToBounds = true
+        backgroundColor = UIColor.clear
+        month.layer.cornerRadius = 8
+        month.layer.masksToBounds = true
+        firstDriver.layer.cornerRadius = 8
```
Comment (kieran-woodrow) [inline F1/F1/Results/AllResultsTableViewCell.swift:34]: nib storyboard

File: F1/F1/Results/RaceResultViewModel.swift [hunk 45]
```diff
@@ -0,0 +1,46 @@
...
+
+import Foundation
+class RaceResultsViewModel {
+    private var repository: ResultsRepositoryType?
+    private weak var delegate: ViewModelDelegate?
+    private var race: Race?
+    private var raceResult: [RacingResult]?
+
+    init(repository: ResultsRepositoryType,
+         delegate: ViewModelDelegate) {
```
[the remark this answers is not in the document]
Comment (kieran-woodrow) [reply F1/F1/Results/RaceResultViewModel.swift:16]: return statements

**Review by kieran-woodrow — CHANGES_REQUESTED** [review]

File: F1/F1/Constructor/constructorTableViewCell.swift (hunk 23 above)
Comment (ksookdhew) [reply F1/F1/Constructor/constructorTableViewCell.swift:37]: Hey, So here im checking that it has at least two not that it is not empty. 😄

**Review by kieran-woodrow — APPROVED** [review]

File: F1/F1/Standings/DriverStandings/DriverStandingsViewController.swift [hunk 46]
```diff
@@ -0,0 +1,33 @@
...
+//
+//  Created by Kaitlyn Sookdhew
+//
+
+import UIKit
+
+class DriverStandingsViewController: UIViewController {
+
+    private lazy var viewModel = DriverStandingsViewModel(repository: DriverStandingsRepository(),
+                                                      delegate: self)
```
Comment (simekanidvt) [inline F1/F1/Standings/DriverStandings/DriverStandingsViewController.swift:13]: This can stay on one line

File: F1/F1/Standings/ConstructorStandings/Repository/ConstructorStandingsRepository.swift [hunk 47]
```diff
@@ -0,0 +1,26 @@
...
+protocol ConstructorStandingsRepositoryType: AnyObject {
+    func fetchConstructorStandingsResults(completion: @escaping(ConstructorStandingsResults))
+}
+
+// MARK: Repository
+class ConstructorStandingsRepository: ConstructorStandingsRepositoryType {
+
+    func fetchConstructorStandingsResults(completion: @escaping (ConstructorStandingsResults)) {
+        URLSession.shared.request(endpoint:
+                                    Endpoints.constructorStanding, method: .GET, completion: completion)
```
Comment (simekanidvt) [inline F1/F1/Standings/ConstructorStandings/Repository/ConstructorStandingsRepository.swift:23]: Check indenting here

File: F1/F1/Standings/ConstructorStandings/ConstructorStandingsViewController.swift [hunk 48]
```diff
@@ -0,0 +1,33 @@
...
+// MARK: - ViewModel Delegate
+
+extension ConstructorStandingsViewController: ViewModelDelegate {
+
+    func reloadView() {
+        // tableView.reloadData()
+    }
+
+    func show(error: String) {
+        // displayAlert(title: "Error", message: error, buttonTitle: "Ok")
```
Comment (simekanidvt) [inline F1/F1/Standings/ConstructorStandings/ConstructorStandingsViewController.swift:31]: Commented code

File: F1/F1/Standings/DriverStandings/Repository/DriverStandingsRepository.swift [hunk 49]
```diff
@@ -0,0 +1,25 @@
...
+protocol DriverStandingsRepositoryType: AnyObject {
+    func fetchDriverStandingsResults(completion: @escaping(DriverStandingsResults))
+}
+
+// MARK: Repository
+class DriverStandingsRepository: DriverStandingsRepositoryType {
+
+    func fetchDriverStandingsResults(completion: @escaping (DriverStandingsResults)) {
+        URLSession.shared.request(endpoint:
+                                    Endpoints.driverStanding, method: .GET, completion: completion)
```
Comment (simekanidvt) [inline F1/F1/Standings/DriverStandings/Repository/DriverStandingsRepository.swift:23]: This can probably be on one line

File: F1/F1/Standings/DriverStandingsTableViewCell.swift [hunk 50]
```diff
@@ -0,0 +1,40 @@
...
+//  Created by Kaitlyn Sookdhew on 2024/04/10.
+//
+
+import UIKit
+
+class DriverStandingsTableViewCell: UITableViewCell {
+
+    @IBOutlet weak var position: UILabel!
+    @IBOutlet weak var name: UILabel!
+    @IBOutlet weak var surname: UILabel!
```
Comment (simekanidvt) [inline F1/F1/Standings/DriverStandingsTableViewCell.swift:14]: Private

**Review by simekanidvt — APPROVED** [review]

File: F1/F1/Login/LoginViewModel.swift [hunk 51]
```diff
@@ -0,0 +1,29 @@
...
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/10.
+//
+
+import Foundation
+
+class LoginViewModel {
+
+    // MARK: Variables
+    private let username = "Admin"
```
Comment (kieran-woodrow) [inline F1/F1/Login/LoginViewModel.swift:13]: You can import coredata or firebase for login instead of hardcoded details

File: F1/F1/Race/RaceScheduleTableViewCell.swift [hunk 52]
```diff
@@ -0,0 +1,29 @@
...
+    // MARK: Functions
+    func populateWith(sessionName: String, sessionDate: DateComponents, sessionTime: String) {
+        raceDay.text = "\(sessionDate.day ?? 0)"
+        raceMonth.text = "\(Constants.monthAbbreviations[(sessionDate.month ?? 1)-1])"
+        session.text = sessionName
+        time.text = sessionTime
+    }
+
+    static func nib() -> UINib {
+        return UINib(nibName: Identifiers.raceScheduleIndentifier, bundle: nil)
```
Comment (kieran-woodrow) [inline F1/F1/Race/RaceScheduleTableViewCell.swift:27]: all of these static func nib() can remove the return. i'm sure it will reduce a few warnings as well

File: F1/F1/Results/RacingResultViewController.swift [hunk 53]
```diff
@@ -0,0 +1,84 @@
...
+
+    // MARK: Variables
+    var race: Race?
+
+    // MARK: Functions
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupTableView()
+        viewModel.setRaceResult(raceResult: race)
+        raceName.text = race?.circuit.location.country
```
Comment (kieran-woodrow) [inline F1/F1/Results/RacingResultViewController.swift:25]: should be done from the viewModel

File: F1/F1/Results/Repository/RacingResults.swift [hunk 54]
```diff
@@ -0,0 +1,102 @@
...
+        case season, round, url, raceName
+        case circuit = "Circuit"
+        case date, time
+        case results = "Results"
+    }
+}
+
+// MARK: - RaceResult
+struct RaceResult: Codable {
+    let number, position, positionText, points: String
```
Comment (kieran-woodrow) [inline F1/F1/Results/Repository/RacingResults.swift:58]: indenting. And please check that all your models follow the same convention. All strings on one line, or all separately. Same for all other types. You cleaned it up in some areas, but in areas like here, it follows a different convention

File: F1/F1/Results/ResultsViewController.swift [hunk 55]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = allResultsTableView
+            .dequeueReusableCell(withIdentifier: Identifiers.allResultsTableViewCell) as? AllResultsTableViewCell else {
+            return UITableViewCell()
+        }
+        guard let result = viewModel.allResult(atIndex: indexPath.section) else { return UITableViewCell() }
+        cell.populateWith(race: result, raceDate: viewModel.allResultDate(result: result))
+        let bgColorView = UIView()
```
Comment (kieran-woodrow) [inline F1/F1/Results/ResultsViewController.swift:74]: is this necessary here? Can you not do it in the storyboard?
or in the cell itself?

File: F1/F1/Results/ResultsViewModel.swift [hunk 56]
```diff
@@ -0,0 +1,81 @@
+//
+//  ResultsViewModel.swift
+//  F1
+//
+//  Created by Kaitlyn Sookdhew on 2024/04/03.
+//
+
+import Foundation
```
Comment (kieran-woodrow) [inline F1/F1/Results/ResultsViewModel.swift:8]: new line after import Foundation

**Review by kieran-woodrow — CHANGES_REQUESTED** [review]
nice job. Just small things.
Only other thing you need to implement is coreData olease (unless i've missed it)
